databricks_cli/dbfs/api.py

```diff
@@ -75,12 +80,29 @@ class DbfsErrorCodes(object):
     RESOURCE_DOES_NOT_EXIST = 'RESOURCE_DOES_NOT_EXIST'
     RESOURCE_ALREADY_EXISTS = 'RESOURCE_ALREADY_EXISTS'
     PARTIAL_DELETE = 'PARTIAL_DELETE'
+    TOO_MANY_REQUESTS = 'TOO_MANY_REQUESTS'
+
+
+def retry_429(func):
+    @retry(wait=wait_random_exponential(multiplier=EXPONENTIAL_BACKOFF_MULTIPLIER, 
+           max=MAX_SECONDS_WAIT), retry=retry_if_exception_type(RateLimitException), 
+           stop=stop_after_attempt(MAX_RETRY_ATTEMPTS))
+    def wrapped_function(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except HTTPError as e:
+            if e.response.status_code == 429:
+                click.echo("Rate limit exceeded. Retrying with exponential backoff.")
```

Comment:
Is it possible to get a retry count or other information about when the next retry will happen?

Reply:
It might be that `after_log` or `before_log` output this kind of info (see [docs](https://tenacity.readthedocs.io/en/latest/#before-and-after-retry-and-logging)). Pushkin, can you check what they output?

Reply:
As @bogdanghita-db suggested, We tried out the various options provided. The `after_log` option on the decorator produces the following output. 
```
Rate limit exceeded. Retrying with exponential backoff.
Rate limit exceeded. Retrying with exponential backoff.
Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api:Finished call to 'databricks_cli.dbfs.api.retry_429.<locals>.wrapped_function' after 5.110(s), this was the 4th time calling it.
WARNING:databricks_cli.dbfs.api:Finished call to 'databricks_cli.dbfs.api.retry_429.<locals>.wrapped_function' after 32.712(s), this was the 5th time calling it.
WARNING:databricks_cli.dbfs.api:Finished call to 'databricks_cli.dbfs.api.retry_429.<locals>.wrapped_function' after 4.487(s), this was the 2nd time calling it.
```

Although, this mentions the time, It's a little confusing as it says `finished call after 4.487(s)` which doesn't mean the function took that long but rather that's how long the wait was. 

The `before_log` option on the other hand don't mention the time at all. Just the attempt number, before attempting as the name suggests. 
```
Rate limit exceeded. Retrying with exponential backoff.
Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api:Starting call to 'databricks_cli.dbfs.api.retry_429.<locals>.wrapped_function', this is the 1st time calling it.
WARNING:databricks_cli.dbfs.api:Starting call to 'databricks_cli.dbfs.api.retry_429.<locals>.wrapped_function', this is the 1st time calling it.
```

Let us know what you think? IMO It might be rather confusing for the user to see these messages.

Reply:
I see in the docs that it's possible to have a custom log like this I think
```
import logging

logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)

logger = logging.getLogger(__name__)

def my_before_sleep(retry_state):
    if retry_state.attempt_number < 1:
        loglevel = logging.INFO
    else:
        loglevel = logging.WARNING
    logger.log(
        loglevel, 'Retrying %s: attempt %s ended with: %s',
        retry_state.fn, retry_state.attempt_number, retry_state.outcome)

@retry(stop=stop_after_attempt(3), before_sleep=my_before_sleep)
def raise_my_exception():
    raise MyException("Fail")

try:
    raise_my_exception()
except RetryError:
    pass
```

Reply:
Yeah, I tried to play with it and got this working. Note these attempts are for different requests. Since they are concurrent, you can't control the order of output ofc. 
```
WARNING:databricks_cli.dbfs.api: Received 429 REQUEST_LIMIT_EXCEEDED for attempt 3. Retrying in 3.3262099571644095 seconds.
WARNING:databricks_cli.dbfs.api: Received 429 REQUEST_LIMIT_EXCEEDED for attempt 2. Retrying in 2.1817190101853114 seconds.
Rate limit exceeded. Retrying with exponential backoff.
Rate limit exceeded. Retrying with exponential backoff.
Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: Received 429 REQUEST_LIMIT_EXCEEDED for attempt 1. Retrying in 0.2708935318360691 seconds.
```

Reply:
Clarified with @gotibhai, the out of order log messages is an artifact of the test script he used. When one uses the CLI by itself, it should print the messages in order.